plugins/scaffolder/api-report.md

```diff
@@ -73,36 +80,78 @@ export const EntityNamePickerFieldExtension: FieldExtensionComponent<
 // @public
 export const EntityPickerFieldExtension: FieldExtensionComponent<
   string,
-  EntityPickerUiOptions
+  {
```

Comment:
I wonder if it's possible to just get this type exported in the `api-report.md` rather than the `z.ZodObject` stuff below? But maybe not?

Reply:
I'm drawing a blank on how to actually get the pure typings to be exported nothing I've tried have worked so far 😕 

I did have one suggestion in my previous [comment](https://github.com/backstage/backstage/pull/14008#issuecomment-1287362789) on one way to keep the api-report clean. Other than that, if you're aware of any typescript/api-extractor magic/hack I can do here, I'm all ears 🙏

Reply:
hmm - ok, wanna bring @Rugvip in here if he thinks there's any solution or whats best moving forward to be honest.

Reply:
I do think we want the options to be part of the API report, since they're part of the public API even if you can't interact with them directly through code. The main option I can think of here to clean things up is to try to wrap things up in our own layer, potentially with some hack to make sure it doesn't get unwrapped. I'll give it a couple of mins to see if I can find some pattern that might work

Reply:
Alright I think this approach could lead to something pretty clean. It's not refined and just a bunch of building blocks for how to actually get something working, needs a bunch of restructuring still 😅 

The TL;DR is that by wrapping things in our own utility function to create schemas we can control what we expose from Zod, and probably avoid Zod exports altogether.

I've got the following as an internal helper:

```ts

function makeSchema<T extends z.ZodType>(
  schema: T,
): {
  uiOptions: JSONSchema7;
  inputType: T extends z.ZodType<any, any, infer I> ? I : never;
} {
  return {
    uiOptions: zodToJsonSchema(schema) as JSONSchema7,
    inputType: null as any,
  };
}
```

It creates the schema and picks out the input type, because I think that's all we're interested in from the Zod schema, right?

It's used like this:

```ts
/**
 * @public
 */
export const OwnerPickerSchema = makeSchema(
  z.object({
    allowedKinds: z
      .array(z.string())
      .default(['Group', 'User'])
      .optional()
      .describe(
        'List of kinds of entities to derive options from. Defaults to Group and User',
      ),
    allowArbitraryValues: z
      .boolean()
      .optional()
      .describe('Whether to allow arbitrary user input. Defaults to true'),
    defaultNamespace: z
      .union([z.string(), z.literal(false)])
      .optional()
      .describe(
        'The default namespace. Options with this namespace will not be prefixed.',
      ),
  }),
);
```

You can pick it apart to export separate things if needed:

```ts
export type OwnerPickerUiOptions = typeof OwnerPickerSchema.inputType;
```

In the API report that shows up like this:

```ts
// @public (undocumented)
export const OwnerPickerSchema: {
  uiOptions: JSONSchema7;
  inputType: {
    defaultNamespace?: string | false | undefined;
    allowedKinds?: string[] | undefined;
    allowArbitraryValues?: boolean | undefined;
  };
};

// @public
export type OwnerPickerUiOptions = typeof OwnerPickerSchema.inputType;
```

Hope that helps! 🤞

Reply:
@Rugvip++! Amazing typescript wizardry 👏  (TIL 😃) 

Updated the change with your suggested solution and it worked like a charm 👌